cpp/src/bitmask/bit_container.cuh

```
@@ -0,0 +1,223 @@
+    bit_container::set_bit_unsafe(valid_, record_idx);
+  }
+
+
+  /**
+   * Clear a bit (not thread-safe)
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __device__ void ClearBitUnsafe(T record_idx) {
+    bit_container::clear_bit_unsafe(record_idx);
+  }
+
+  /**
+   * Set a bit in a thread-safe manner
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __device__ void SetBit(T record_idx) {
+    int rec = bit_container::which_word(record_idx);
+    int bit = bit_container::which_bit(record_idx);
+
+    atomicOr( &valid_[rec], (1U << bit));
+  }
+
+
+  /**
+   * Clear a bit in a thread-safe manner
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __device__ void ClearBit(T record_idx) {
+    int rec = bit_container::which_word(record_idx);
+    int bit = bit_container::which_bit(record_idx);
+
+    atomicAnd( &valid_[rec], ~(1U << bit));
+  }
+
+  /**
+   * Get the number of words in this bit container
+   *
+   * @return the number of words
+   */
+  __device__ int NumWords() const {
+    return bit_container::num_words(bitlength_);
+  }
+
+  /**
+   * Getter for valid
+   *
+   * @return pointer to valid bit container
+   */
+  __host__ __device__ bit_container_t *GetValid() const {
+    return valid_;
+  }
+
+  /**
+   * Get length
+   *
+   * @return length of bit container in bits
+   */
+  __host__ __device__ int Length() const {
+    return bitlength_;
+  }
+
+
+  /**
+   * Allocate device space for the valid bitmap.
+   *
+   * @param[in]  number_of_records     number of records
+   * @param[in]  fill_value            optional, should the memory be initialized to all 0 or 1s. All other values indicate un-initialized
+   * @return pointer to allocated space on the device
+   */
+  template <typename T>
+  __host__
+  static bit_container_t * CreateBitContainer(T number_of_records, int fill_value = -1) {
```

This should be a free function in the other file rather than a `static` member function.
This also needs to accept a parameter for the padding that should be defaulted to 64 bytes (per Arrow spec requirements).

---

The padding parameter would specify that the buffer allocation for the bitmasks be a multiple of the specified padding. I.e., for the default of 64 bytes, all allocations for the bit masks will be a multiple of 64 bytes. 

I believe that this formula will give you what you need. Note that this will give an `allocation_size == padding` when `number_of_records == 0`, so probably need to add a check if `number_of_records == 0`, return an error.

```
gdf_size_type allocation_size = number_of_records - 1 - (number_of_records - 1) % padding + padding ;
```